backend/service/aws/cache.go

```diff
@@ -0,0 +1,166 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/golang/protobuf/ptypes"
+	"go.uber.org/zap"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const topologyObjectChanBufferSize = 5000
+const topologyLockId = 1
+
+func (c *client) CacheEnabled() bool {
+	return true
+}
+
+func (c *client) StartTopologyCaching(ctx context.Context) (<-chan *topologyv1.UpdateCacheRequest, error) {
+	if !c.topologyLock.TryAcquire(topologyLockId) {
+		return nil, errors.New("TopologyCahing is already in progress")
+	}
+
+	go func() {
+		<-ctx.Done()
+		c.topologyLock.Release(topologyLockId)
+	}()
+
+	go c.processRegionTopologyObjects(ctx)
+	return c.topologyObjectChan, nil
+}
+
+func (c *client) processRegionTopologyObjects(ctx context.Context) {
+	ticker := time.NewTicker(time.Minute * 10)
+	for ; true; <-ticker.C {
+		wg := &sync.WaitGroup{}
+		for name, client := range c.clients {
+			c.log.Info("processing topology objects for region", zap.String("region", name))
+			go c.processAllEC2Instances(ctx, client, wg)
+			go c.processAllAutoScalingGroups(ctx, client, wg)
+			go c.processAllKinesisStreams(ctx, client, wg)
+		}
+		wg.Wait()
```

Comment:
i wonder if this is the right approach. do we not want to refresh kinesis streams if instances are taking longer than a tick to complete? with 10 minutes, sure probably not a concern, but what if we want to lower the refresh rate?

Reply:
Yeah good question I think I like refactoring this into each resource has its own ticker so a user could more easily control the rate at which we poll for individual resource. I can also leave a comment to make this configurable in the future as well.